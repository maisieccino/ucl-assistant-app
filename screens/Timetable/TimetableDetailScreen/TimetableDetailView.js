import PropTypes from "prop-types"
import React from "react"
import { StyleSheet, View } from "react-native"
import MapView from "react-native-maps"

import Button from "../../../components/Button"
import { Page } from "../../../components/Containers"
import {
  BodyText,
  ErrorText,
  Link,
  SubtitleText,
  TitleText,
} from "../../../components/Typography"
import { LocalisationManager, MailManager, MapsManager } from "../../../lib"
import MapStyle from "../../../styles/Map"

const styles = StyleSheet.create({
  contactPerson: {
    marginBottom: 20,
    marginTop: 20,
  },
  emailButton: {
    marginTop: 10,
  },
})

class TimetableDetailView extends React.Component {
  static propTypes = {
    contact: PropTypes.string,
    date: PropTypes.string,
    end_time: PropTypes.string,
    initialRegion: PropTypes.shape({
      latitude: PropTypes.number,
      latitudeDelta: PropTypes.number,
      longitude: PropTypes.number,
      longitudeDelta: PropTypes.number,
    }),
    location: PropTypes.shape({
      address: PropTypes.arrayOf(PropTypes.string),
      coordinates: PropTypes.shape({
        lat: PropTypes.string,
        lng: PropTypes.string,
      }),
      name: PropTypes.string,
      type: PropTypes.string,
    }),
    module: PropTypes.shape({
      department_name: PropTypes.string,
      email: PropTypes.string,
      lecturer: PropTypes.shape({
        department_id: PropTypes.string,
        department_name: PropTypes.string,
        email: PropTypes.string,
        name: PropTypes.string,
      }),
      name: PropTypes.string,
    }),
    navigation: PropTypes.shape().isRequired,
    session_group: PropTypes.string,
    session_type_str: PropTypes.string,
    start_time: PropTypes.string,
  }

  static defaultProps = {
    contact: ``,
    date: `2019-01-01`,
    end_time: ``,
    initialRegion: {
      latitude: 51.5246586,
      latitudeDelta: 0.0012,
      longitude: -0.1339784,
      longitudeDelta: 0.0071,
    },
    location: {
      address: [],
      coordinates: {
        lat: `51.5246586`,
        lng: `-0.1339784`,
      },
      name: ``,
      type: `DB`,
    },
    module: {
      department_name: ``,
      email: ``,
      lecturer: {
        department_id: ``,
        department_name: ``,
        email: ``,
        name: ``,
      },
      name: ``,
    },
    session_group: ``,
    session_type_str: ``,
    start_time: ``,
  }

  navigateToLocation = ({ lat, lng, address }) => () => {
    if (lat && lng) {
      MapsManager.navigateToCoords({ lat, lng })
    } else {
      MapsManager.navigateToAddress(address.join())
    }
  }

  sendEmail = (email) => () => MailManager.composeAsync({
    recipients: [email],
  })

  openRoomSearch = (roomName) => () => {
    const { navigation } = this.props
    navigation.navigate(`Main`, {
      params: {
        params: { query: roomName },
        screen: `RoomsSearch`,
      },
    })
  }

  renderContactPerson = ({
    contactTypeStr,
    contactPerson,
    departmentName,
    email,
  }) => {
    const validContactPerson = contactPerson && contactPerson.length > 0
    const validDepartment = departmentName && departmentName.length > 0
    const validEmail = email && email.length > 0
    if (!validContactPerson && !validEmail) {
      return null
    }
    return (
      <View style={styles.contactPerson}>
        <SubtitleText>{contactTypeStr}</SubtitleText>
        {validContactPerson ? (
          <BodyText>
            {`${contactPerson} - ${validDepartment ? departmentName : null}`}
          </BodyText>
        ) : null}
        {validEmail ? (
          <Button onPress={this.sendEmail(email)} style={styles.emailButton}>
            Send Email
          </Button>
        ) : null}
      </View>
    )
  }

  render() {
    let contactTypeStr = ``
    const {
      date,
      location,
      initialRegion,
      contact: contactPerson,
      module: { name: moduleName, department_name: departmentName, email },
      session_type_str: sessionTypeStr,
      session_group: sessionGroup,
      start_time: startTime,
      end_time: endTime,
    } = this.props
    const sessionType = sessionTypeStr.toLowerCase()
    switch (true) {
      case sessionType === `lecture`:
      case sessionType === `seminar`: {
        contactTypeStr = `Lecturer`
        break
      }
      case sessionType === `practical`:
      case sessionType === `problem based learning`: {
        contactTypeStr = `Instructor`
        break
      }
      default: {
        contactTypeStr = `Contact`
      }
    }

    const { lat, lng } = location.coordinates
    const latitude = parseFloat(lat, 10) || initialRegion.latitude
    const longitude = parseFloat(lng, 10) || initialRegion.longitude
    const { address, type: locationType, name: locationName } = location

    return (
      <Page>
        <TitleText>{moduleName}</TitleText>
        <BodyText>{LocalisationManager.parseToMoment(date).format(`dddd, Do MMMM YYYY`)}</BodyText>
        <BodyText>
          {`${startTime} - ${endTime}`}
        </BodyText>
        {locationType === `CB` ? (
          <Link onPress={this.openRoomSearch(locationName)}>
            {locationName}
          </Link>
        ) : (
            <BodyText>{locationName}</BodyText>
        )}
        <BodyText>
          {`Type: ${sessionTypeStr}`}
        </BodyText>
        {sessionGroup && sessionGroup.length > 0 ? (
          <BodyText>
            {`Group ${sessionGroup}`}
          </BodyText>
        ) : null}
        {(!lat || !lng) && (
          <ErrorText>
            Error: We couldn&apos;t fetch coordinates for this venue,
            &nbsp;so the map may be incorrect.
          </ErrorText>
        )}
        <MapView
          style={MapStyle.wideMap}
          initialRegion={initialRegion}
          region={{
            latitude,
            latitudeDelta: initialRegion.latitudeDelta,
            longitude,
            longitudeDelta: initialRegion.longitudeDelta,
          }}
        >
          <MapView.Marker coordinate={{ latitude, longitude }} />
        </MapView>
        <Button onPress={this.navigateToLocation({ address, lat, lng })}>
          Directions
        </Button>

        {this.renderContactPerson({
          contactPerson, contactTypeStr, departmentName, email,
        })}
      </Page>
    )
  }
}

export default TimetableDetailView
