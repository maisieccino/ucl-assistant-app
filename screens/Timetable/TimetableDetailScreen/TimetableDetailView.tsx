import React from "react"
import { StyleSheet, View } from "react-native"
import MapView, { Marker } from "react-native-maps"

import Button from "../../../components/Button"
import { Page } from "../../../components/Containers"
import {
  BodyText,
  ErrorText,
  Link,
  SubtitleText,
  TitleText,
} from "../../../components/Typography"
import { LocalisationManager, MailManager, MapsManager } from "../../../lib"
import type {
  MailManagerComposeAsyncReturnType,
} from "../../../lib/MailManager"
import MapStyle from "../../../styles/Map"
import type { Region, TimetableEvent } from "../../../types/uclapi"
import type { TimetableNavigationType } from "../TimetableNavigator"

const styles = StyleSheet.create({
  contactPerson: {
    marginBottom: 20,
    marginTop: 20,
  },
  emailButton: {
    marginTop: 10,
  },
})

interface Props extends TimetableEvent {
  navigation: TimetableNavigationType,
  initialRegion: Region,
}


class TimetableDetailView extends React.Component<Props> {
  navigateToLocation = (
    {
      lat,
      lng,
      address,
    }: {
      lat: string,
      lng: string,
      address: Array<string>,
    },
  ) => (): void => {
    if (lat && lng) {
      MapsManager.navigateToCoords({ lat, lng })
    } else {
      MapsManager.navigateToAddress(address.join())
    }
  }

  sendEmail = (
    email: string,
  ) => (): MailManagerComposeAsyncReturnType => MailManager.composeAsync({
    recipients: [email],
  })

  openRoomSearch = (roomName: string) => (): void => {
    const { navigation } = this.props
    navigation.navigate(`Main`, {
      params: {
        initial: false,
        params: { query: roomName },
        screen: `RoomsSearch`,
      },
      screen: `Rooms`,
    })
  }

  renderContactPerson = ({
    contactTypeStr,
    contactPerson,
    departmentName,
    email,
  }: {
    contactTypeStr: string,
    contactPerson: string,
    departmentName: string,
    email: string,
  }): React.ReactNode => {
    const validContactPerson = contactPerson && contactPerson.length > 0
    const validDepartment = departmentName && departmentName.length > 0
    const validEmail = email && email.length > 0
    if (!validContactPerson && !validEmail) {
      return null
    }
    return (
      <View style={styles.contactPerson}>
        <SubtitleText>{contactTypeStr}</SubtitleText>
        {validContactPerson ? (
          <BodyText>
            {`${contactPerson} - ${validDepartment ? departmentName : null}`}
          </BodyText>
        ) : null}
        {validEmail ? (
          <Button onPress={this.sendEmail(email)} style={styles.emailButton}>
            Send Email
          </Button>
        ) : null}
      </View>
    )
  }

  render(): React.ReactElement {
    let contactTypeStr = ``
    const {
      date = `2019-01-01`,
      location = {
        address: [],
        coordinates: {
          lat: `0`,
          lng: `0`,
        },
        name: ``,
        type: ``,
      },
      initialRegion = {
        latitude: 51.5246586,
        latitudeDelta: 0.0012,
        longitude: -0.1339784,
        longitudeDelta: 0.0071,
      },
      contact: contactPerson = ``,
      module: { name: moduleName, department_name: departmentName, email } = {},
      session_type_str: sessionTypeStr = ``,
      session_group: sessionGroup = ``,
      start_time: startTime = ``,
      end_time: endTime = ``,
    } = this.props
    const sessionType = sessionTypeStr.toLowerCase()
    switch (true) {
      case sessionType === `lecture`:
      case sessionType === `seminar`: {
        contactTypeStr = `Lecturer`
        break
      }
      case sessionType === `practical`:
      case sessionType === `problem based learning`: {
        contactTypeStr = `Instructor`
        break
      }
      default: {
        contactTypeStr = `Contact`
      }
    }

    const { lat, lng } = location.coordinates
    const latitude = parseFloat(lat) || initialRegion.latitude
    const longitude = parseFloat(lng) || initialRegion.longitude
    const { address, type: locationType, name: locationName } = location

    return (
      <Page>
        <TitleText>{moduleName}</TitleText>
        <BodyText>
          {LocalisationManager.parseToMoment(date).format(`dddd, Do MMMM YYYY`)}
        </BodyText>
        <BodyText>
          {`${startTime} - ${endTime}`}
        </BodyText>
        {locationType === `CB` ? (
          <Link onPress={this.openRoomSearch(locationName)}>
            {locationName}
          </Link>
        ) : (
            <BodyText>{locationName}</BodyText>
        )}
        <BodyText>
          {`Type: ${sessionTypeStr}`}
        </BodyText>
        {sessionGroup && sessionGroup.length > 0 ? (
          <BodyText>
            {`Group ${sessionGroup}`}
          </BodyText>
        ) : null}
        {(!lat || !lng) && (
          <ErrorText>
            Error: We couldn&apos;t fetch coordinates for this venue,
            &nbsp;so the map may be incorrect.
          </ErrorText>
        )}
        <MapView
          style={MapStyle.wideMap}
          initialRegion={initialRegion}
          region={{
            latitude,
            latitudeDelta: initialRegion.latitudeDelta,
            longitude,
            longitudeDelta: initialRegion.longitudeDelta,
          }}
        >
          <Marker coordinate={{ latitude, longitude }} />
        </MapView>
        <Button onPress={this.navigateToLocation({ address, lat, lng })}>
          Directions
        </Button>

        {this.renderContactPerson({
          contactPerson, contactTypeStr, departmentName, email,
        })}
      </Page>
    )
  }
}

export default TimetableDetailView
