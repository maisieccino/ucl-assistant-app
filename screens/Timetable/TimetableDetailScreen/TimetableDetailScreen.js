import PropTypes from "prop-types"
import React, { Component } from "react"
import { connect } from "react-redux"

import { timetableSelector } from "../../../selectors/timetableSelectors"
import TimetableDetailView from "./TimetableDetailView"

export class TimetableDetailScreen extends Component {
  static navigationOptions = ({ route }) => ({
    title: `Event: ${route.params.code}`,
  })

  static mapStateToProps = (state) => ({
    timetable: timetableSelector(state),
  })

  static propTypes = {
    navigation: PropTypes.shape().isRequired,
    route: PropTypes.shape().isRequired,
    timetable: PropTypes.shape(),
  }

  static defaultProps = {
    timetable: {},
  }

  constructor(props) {
    super(props)
    const { params } = props.route
    const { date = `2018-01-01`, time, code } = params
    this.state = {
      // pre-defined
      code,
      date,
      event: {},
      // from timetable
      time,
    }
  }

  componentDidMount() {
    this.timetableLoadedTest()
  }

  timetableLoadedTest = () => {
    const { timetable } = this.props
    if (Object.keys(timetable).length > 0) {
      this.findEvent()
    }
  }

  async findEvent() {
    const { date, time, code } = this.state
    const { timetable } = this.props
    const timetableDay = (timetable[date] || {}).timetable || []
    const event = timetableDay.filter(
      (ev) => ev.module.module_id === code && ev.start_time === time,
    )[0]
    await this.setState({ event })
  }

  render() {
    const initialRegion = {
      latitude: 51.5246586,
      latitudeDelta: 0.0012,
      longitude: -0.1339784,
      longitudeDelta: 0.0071,
    }
    const { navigation } = this.props
    const { date, event } = this.state
    return (
      <TimetableDetailView
        initialRegion={initialRegion}
        date={date}
        {...event}
        navigation={navigation}
      />
    )
  }
}

export default connect(TimetableDetailScreen.mapStateToProps)(
  TimetableDetailScreen,
)
