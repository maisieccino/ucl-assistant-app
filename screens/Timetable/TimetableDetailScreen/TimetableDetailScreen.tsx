import type { BottomTabNavigationProp } from "@react-navigation/bottom-tabs"
import { CompositeNavigationProp, RouteProp } from "@react-navigation/native"
import type { StackNavigationProp } from "@react-navigation/stack"
import React from "react"
import { connect, ConnectedProps } from "react-redux"

import type { AppStateType } from "../../../configureStore"
import type {
  MainTabNavigatorParamList,
} from "../../../navigation/MainTabNavigator"
import type { RootStackParamList } from "../../../navigation/RootNavigation"
import { timetableSelector } from "../../../selectors/timetableSelectors"
import type { TimetableNavigatorParamList } from '../TimetableNavigator'
import TimetableDetailView from "./TimetableDetailView"

// eslint-disable-next-line quotes
type RouteType = RouteProp<TimetableNavigatorParamList, 'TimetableDetail'>

interface Props extends PropsFromRedux {
  navigation: CompositeNavigationProp<
    StackNavigationProp<TimetableNavigatorParamList>,
    CompositeNavigationProp<
      BottomTabNavigationProp<MainTabNavigatorParamList>,
      StackNavigationProp<RootStackParamList>
    >
  >,
  route: RouteType,
}

interface State {
  date: string,
  time: string,
  code: string,
  event: unknown,
}

export class TimetableDetailScreen extends React.Component<Props, State> {
  static navigationOptions = ({ route }: { route: RouteType, }) => ({
    title: `Event: ${route.params.code}`,
  })

  constructor(props: Props) {
    super(props)
    const { params } = props.route
    const { date = `2018-01-01`, time, code } = params
    this.state = {
      // pre-defined
      code,
      date,
      event: {},
      // from timetable
      time,
    }
  }

  componentDidMount(): void {
    this.timetableLoadedTest()
  }

  timetableLoadedTest = (): void => {
    const { timetable } = this.props
    if (Object.keys(timetable).length > 0) {
      this.findEvent()
    }
  }

  async findEvent(): Promise<void> {
    const { date, time, code } = this.state
    const { timetable } = this.props
    const timetableDay = (timetable[date] || {}).timetable || []
    const event = timetableDay.filter(
      (ev) => ev.module.module_id === code && ev.start_time === time,
    )[0]
    await this.setState({ event })
  }

  render(): React.ReactElement {
    const initialRegion = {
      latitude: 51.5246586,
      latitudeDelta: 0.0012,
      longitude: -0.1339784,
      longitudeDelta: 0.0071,
    }
    const { navigation } = this.props
    const { date, event } = this.state
    return (
      <TimetableDetailView
        initialRegion={initialRegion}
        date={date}
        {...event}
        navigation={navigation}
      />
    )
  }
}

const connector = connect(
  (state: AppStateType) => ({
    timetable: timetableSelector(state),
  }),
)

type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(TimetableDetailScreen)
