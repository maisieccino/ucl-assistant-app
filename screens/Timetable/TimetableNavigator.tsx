import {
  createStackNavigator,
} from '@react-navigation/stack'
import React from 'react'

import TimetableDetailScreen from "./TimetableDetailScreen"
import TimetableScreen from "./TimetableScreen"

type TimetableNavigatorParamList = {
  Timetable: undefined,
  TimetableDetail: {
    code: string,
    date: string,
    module: string,
    time: string,
  },
}

const Stack = createStackNavigator<TimetableNavigatorParamList>()

const TimetableNavigator = (): React.ReactElement => (
  <Stack.Navigator initialRouteName="Timetable">
    <Stack.Screen
      options={{ headerShown: false }}
      name="Timetable"
      component={TimetableScreen}
    />
    <Stack.Screen
      name="TimetableDetail"
      component={TimetableDetailScreen}
      options={{ title: `` }}
    />
  </Stack.Navigator>
)

export default TimetableNavigator
