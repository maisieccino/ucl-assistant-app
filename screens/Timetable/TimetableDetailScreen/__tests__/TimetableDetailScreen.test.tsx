import { RouteProp } from "@react-navigation/native"
/**
 * @jest-environment jsdom
 */

import { cleanup, render } from "@testing-library/react-native"
import React from 'react'
import { TimetableNavigatorParamList } from "../../TimetableNavigator"
import { TimetableDetailScreen } from '../TimetableDetailScreen'

jest.mock(`../TimetableDetailView`, () => (props) => JSON.stringify(props))

describe(`TimetableDetailScreen`, () => {
  let wrapper

  const mockProps = {
    dispatch: jest.fn(),
    navigation: {
      dispatch: jest.fn(),
      navigate: jest.fn(),
    } as any,
    route: {
      params: {
        code: `ABC123`,
        date: `2020-01-01`,
        time: `00:00`,
      },
    // eslint-disable-next-line quotes
    } as RouteProp<TimetableNavigatorParamList, "TimetableDetail">,
    timetable: {
      "2020-01-01": {
        timetable: [
          {
            contact: `Anne Oldman`,
            end_time: `00:01`,
            location: {
              address: [
                `221C Banana Bread Street`,
              ],
              name: `Waffle House`,
            },
            module: {
              department_name: `Criminology`,
              lecturer: {
                department_id: `CRIM`,
                name: `Jack Cloth`,
              },
              module_id: `ABC123`,
            },
            start_time: `00:00`,
          },
        ],
      },
    },
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })
  afterEach(() => {
    cleanup()
  })

  it(`renders the TimetableDetailScreen`, async () => {
    wrapper = await render(<TimetableDetailScreen {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })
})
