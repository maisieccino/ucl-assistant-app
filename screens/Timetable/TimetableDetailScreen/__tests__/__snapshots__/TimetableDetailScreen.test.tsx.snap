// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimetableDetailScreen renders the TimetableDetailScreen 1`] = `"{\\"initialRegion\\":{\\"latitude\\":51.5246586,\\"latitudeDelta\\":0.0012,\\"longitude\\":-0.1339784,\\"longitudeDelta\\":0.0071},\\"date\\":\\"2020-01-01\\",\\"contact\\":\\"Anne Oldman\\",\\"end_time\\":\\"00:01\\",\\"location\\":{\\"address\\":[\\"221C Banana Bread Street\\"],\\"name\\":\\"Waffle House\\"},\\"module\\":{\\"department_name\\":\\"Criminology\\",\\"lecturer\\":{\\"department_id\\":\\"CRIM\\",\\"name\\":\\"Jack Cloth\\"},\\"module_id\\":\\"ABC123\\"},\\"start_time\\":\\"00:00\\",\\"navigation\\":{}}"`;
